docs/high-performance/ParallelHelper.md

```diff
@@ -0,0 +1,117 @@
+
+To introduce the concept of closures, suppose we want to multiply the array elements by a value that is specified at runtime. To do so, we need to "capture" that value in our callback `struct` type. We can do that like so:
+
+```csharp
+public readonly struct ItemsMultiplier : IRefAction<float>
+{
+    private readonly float factor;
+    
+    public ItemsMultiplier(float factor)
+    {
+        this.factor = factor;
+    }
+
+    public void Invoke(ref float x) => x *= this.factor;
+}
+
+// ...
+
+ParallelHelper.ForEach(array, new ItemsMultiplier(3.14f));
+```
+
+We can see that the `struct` now contains a field that represents the factor we want to use to multiply elements, instead of using a constant. And when invoking `ForEach`, we're explicitly creating an instance of our callback type, with the factor we're interested in. Furthermore, in this case the C# compiler is also able to automatically recognize the type arguments we're using, so we can omit them together from the method invocation.
+
+This approach of creating fields for values we need to access from a callback lets us explicitly declare what values we want to capture, which helps makes the code more expressive. This is exactly the same thing that the C# compiler does behind the scenes when we declare a lambda function or local function that accesses some local variable as well.
+
+Here is another example, this time using the `For` API to initialize all the items of an array in parallel. Note how this time we're capturing the target array directly, and we're using the `IAction` `interface` for our callback, which gives our method the current parallel iteration index as argument:
+
+```csharp
+public readonly struct ArrayInitializer : IAction
+{
+    private int[] array;
+    
+    public ArrayInitializer(int[] array)
+    {
+        this.array = array;
+    }
+    
+    public void Invoke(int i)
+    {
+    	this.array[i] = i;
+    }
+}
+
+// ...
+
+ParallelHelper.For(0, array.Length, new ArrayInitializer(array));
+```
+
+**NOTE:** since the callback types are `struct`-s, they're passed _by copy_ to each thread running parallel, not by reference. This means that value types being stored as fields in a callback types will be copied as well. A good practice to remember that detail and avoid errors is to mark the callback `struct` as `readonly`, so that the C# compiler will not let us modify the values of its fields. This only applies to _instance_ fields of a value type: if a callback `struct` has a `static` field of any type, or a reference field, then that value will correctly be shared between parallel threads.
```

Comment:
Use the DocFX note format:

```suggestion
> [!NOTE]
> Since the callback types are `struct`-s, they're passed _by copy_ to each thread running parallel, not by reference. This means that value types being stored as fields in a callback types will be copied as well. A good practice to remember that detail and avoid errors is to mark the callback `struct` as `readonly`, so that the C# compiler will not let us modify the values of its fields. This only applies to _instance_ fields of a value type: if a callback `struct` has a `static` field of any type, or a reference field, then that value will correctly be shared between parallel threads.
```